### PR TITLE
feat: Application Emojis

### PR DESCRIPTION
**Pull request by Lulalaby**
# Description

This implements [Application Emojis](https://discord.com/developers/docs/change-log#application-emoji) (THANKS MAISY).
Breaking change in `DiscordEmoji.TryFromName` (additional required parameter)

## Added methods, properties & objects
- `BaseDiscordClient.Emojis`-  (`IReadOnlyDictionary<ulong, DiscordApplicationEmoji>`)
- `DiscordClient.Emojis`-  (`IReadOnlyDictionary<ulong, DiscordApplicationEmoji>`)
- `DiscordClient.GetApplicationEmojisAsync(bool fetch = false)` - (`Task<IReadOnlyList<DiscordApplicationEmoji>>`)
- `DiscordClient.GetApplicationEmojiAsync(ulong id, bool fetch = false)` - (`Task<DiscordApplicationEmoji?>`)
- `DiscordClient.CreateApplicationEmojiAsync(string name, Stream image)` - (`Task<DiscordApplicationEmoji>`)
- `DiscordClient.ModifyApplicationEmojiAsync(ulong id, string name)` - (`Task<DiscordApplicationEmoji>`)
- `DiscordClient.DeleteApplicationEmojiAsync(ulong id)`
- `DiscordConfiguration.AutoFetchApplicationEmojis`
- `DiscordApplicationEmoji`
- `DiscordApplicationEmoji.ModifyAsync(string name)` - (`Task<DiscordApplicationEmoji>`)
- `DiscordApplicationEmoji.DeleteAsync()`
- `DiscordEmoji.FromApplicationEmote(BaseDiscordClient client, ulong id)`- (`DiscordEmoji`), can be casted to `DiscordApplicationEmoji`

## Changed methods, properties & objects
- `DiscordEmoji.RolesInternal` - Will default to `[]` meaning `RolesInternal.Roles` should never throw an NRE. Made it internal
- `DiscordEmoji.FromName(BaseDiscordClient client, string name, bool includeGuilds = true, bool includeApplication = true)` - (`DiscordEmoji`) , can be casted to `DiscordUnicodeEmoji`, `DiscordGuildEmoji` & `DiscordApplicationEmoji`
- ⚠️ `DiscordEmoji.TryFromName(BaseDiscordClient client, string name, bool includeGuilds, bool includeApplication, out DiscordEmoji emoji)` - (`DiscordEmoji`) , can be casted to `DiscordUnicodeEmoji`, `DiscordGuildEmoji` & `DiscordApplicationEmoji` |  **Breaking:** Additional required parameter `bool includeApplication`

## Additional
- Updated incorrect internal documentation

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

@Aiko-IT-Systems/discatsharp

### CODE REVIEW

File: DisCatSharp/Clients/DiscordClient.cs [hunk 1]
```diff
@@ -615,6 +633,60 @@ public async Task RemoveGuildApplicationCommandsAsync(ulong guildId)
...
 		=> await this.RemoveGuildApplicationCommandsAsync(guild.Id).ConfigureAwait(false);
 
+	/// <summary>
+	/// Gets the application emojis.
+	/// <param name="fetch">Whether to ignore the cache. Defaults to false.</param>
+	/// </summary>
+	/// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+	/// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+	public async Task<IReadOnlyList<DiscordApplicationEmoji>> GetApplicationEmojisAsync(bool fetch = false)
+		=> (fetch ? null : this.InternalGetCachedApplicationEmojis()) ?? await this.ApiClient.GetApplicationEmojisAsync(this.CurrentApplication.Id).ConfigureAwait(false);
```
Comment (TheXorog) [inline DisCatSharp/Clients/DiscordClient.cs:643]: This is rather confusing to read imo. I think we should fall back to how the other if checks are done.

File: DisCatSharp/Clients/DiscordClient.cs [hunk 2]
```diff
@@ -615,6 +633,60 @@ public async Task RemoveGuildApplicationCommandsAsync(ulong guildId)
...
+	/// <summary>
+	/// Gets an application emoji.
+	/// </summary>
+	/// <param name="id">The emoji id.</param>
+	/// <param name="fetch">Whether to ignore the cache. Defaults to false.</param>
+	/// <exception cref="NotFoundException">Thrown when the emoji does not exist.</exception>
+	/// <exception cref="BadRequestException">Thrown when an invalid parameter was provided.</exception>
+	/// <exception cref="ServerErrorException">Thrown when Discord is unable to process the request.</exception>
+	public async Task<DiscordApplicationEmoji?> GetApplicationEmojiAsync(ulong id, bool fetch = false)
+		=> (fetch ? null : this.InternalGetCachedApplicationEmoji(id)) ?? await this.ApiClient.GetApplicationEmojiAsync(this.CurrentApplication.Id, id).ConfigureAwait(false);
```
Comment (TheXorog) [inline DisCatSharp/Clients/DiscordClient.cs:654]: Same here

**Review by TheXorog — CHANGES_REQUESTED** [review]
Rather minor nitpick

File: DisCatSharp/Clients/DiscordClient.cs (hunk 1 above)
Comment (Lulalaby) [reply DisCatSharp/Clients/DiscordClient.cs:643]: This is the usual way for cache-aware methods tho as shorthand, I see no need to change that

File: DisCatSharp/Clients/DiscordClient.cs [hunk 3]
```diff
@@ -1286,29 +1358,44 @@ public async Task<DiscordApplicationCommand> EditGuildApplicationCommandAsync(ul
...
+	/// <returns>The requested emoji.</returns>
+	internal DiscordApplicationEmoji? InternalGetCachedApplicationEmoji(ulong emojiId)
+		=> this.EmojisInternal is null || this.EmojisInternal.Count is 0 ? null : this.EmojisInternal.GetValueOrDefault(emojiId);
+
+	/// <summary>
+	/// Gets the internal cached emojis.
+	/// </summary>
+	/// <returns>The requested emoji.</returns>
+	internal IReadOnlyList<DiscordApplicationEmoji> InternalGetCachedApplicationEmojis()
+		=> this.EmojisInternal is null || this.EmojisInternal.Count is 0 ? [] : this.EmojisInternal.Values.ToList();
```
Comment (Lulalaby) [inline DisCatSharp/Clients/DiscordClient.cs:1386]:
```suggestion
		=> this.EmojisInternal is null || this.EmojisInternal.Count is 0 ? null : this.EmojisInternal.Values.ToList();
```

File: DisCatSharp/Clients/DiscordClient.cs [hunk 4]
```diff
@@ -1286,29 +1358,44 @@ public async Task<DiscordApplicationCommand> EditGuildApplicationCommandAsync(ul
...
+	/// <param name="emojiId">The target emoji id.</param>
+	/// <returns>The requested emoji.</returns>
+	internal DiscordApplicationEmoji? InternalGetCachedApplicationEmoji(ulong emojiId)
+		=> this.EmojisInternal is null || this.EmojisInternal.Count is 0 ? null : this.EmojisInternal.GetValueOrDefault(emojiId);
+
+	/// <summary>
+	/// Gets the internal cached emojis.
+	/// </summary>
+	/// <returns>The requested emoji.</returns>
+	internal IReadOnlyList<DiscordApplicationEmoji> InternalGetCachedApplicationEmojis()
```
Comment (Lulalaby) [inline DisCatSharp/Clients/DiscordClient.cs:1385]:
```suggestion
	internal IReadOnlyList<DiscordApplicationEmoji>? InternalGetCachedApplicationEmojis()
```

**Pull request merged**